### DISCUSSION TITLE
Interaction has already been acknowledged

### DISCUSSION THREAD

**Discussion by ctrl-Felix** [Rapptz/discord.py#7876] [Q&A]
I have a button in a view created with this:
```python
view = discord.ui.View()
item = discord.ui.Button(style=discord.ButtonStyle.blurple, label="Click", custom_id=f"id_{str(id)}")
view.add_item(item=item)
``` 
I then handle the interaction with:
```python
@bot.listen()
async def on_interaction(interaction):
    await interaction.response.defer(ephemeral=True, thinking=True)
    if interaction.user.bot:
        return
    ...
```
But I keep getting this on new views:

```Traceback (most recent call last):
  File "/root/discordtipbotcode/venv/lib/python3.9/site-packages/discord/client.py", line 384, in _run_event
    await coro(*args, **kwargs)
  File "/root/discordtipbotcode/bot/modules/wallet.py", line 408, in on_interaction
    await interaction.response.defer(ephemeral=True, thinking=True)
  File "/root/discordtipbotcode/venv/lib/python3.9/site-packages/discord/interactions.py", line 486, in defer
    await adapter.create_interaction_response(parent.id, parent.token, session=parent._session, params=params)
  File "/root/discordtipbotcode/venv/lib/python3.9/site-packages/discord/webhook/async_.py", line 194, in request
    raise HTTPException(response, data)
discord.errors.HTTPException: 400 Bad Request (error code: 40060): Interaction has already been acknowledged.
Ignoring exception in on_interaction
Traceback (most recent call last):
  File "/root/discordtipbotcode/venv/lib/python3.9/site-packages/discord/client.py", line 384, in _run_event
    await coro(*args, **kwargs)
  File "/root/discordtipbotcode/bot/modules/wallet.py", line 408, in on_interaction
    await interaction.response.defer(ephemeral=True, thinking=True)
  File "/root/discordtipbotcode/venv/lib/python3.9/site-packages/discord/interactions.py", line 486, in defer
    await adapter.create_interaction_response(parent.id, parent.token, session=parent._session, params=params)
  File "/root/discordtipbotcode/venv/lib/python3.9/site-packages/discord/webhook/async_.py", line 194, in request
    raise HTTPException(response, data)
discord.errors.HTTPException: 400 Bad Request (error code: 40060): Interaction has already been acknowledged.
```

Once I restart the bot the older views work. But if I create a new one it's not working...
The error already appears multiple times on the same instance

Addition:
If I add a `asyncio.sleep(2)` before .defer() I get `discord.errors.InteractionResponded: This interaction has already been responded to before`

**Comment by Rapptz** [comment]
There are multiple issues here.

1. You are setting a custom_id explicitly, which might have a chance of conflict unless you're going out of way to ensure they are unique.
2. You are responding to *every* interaction the bot sees instantly with a defer. Any response you do after that is bound to fail.

You should use the callback attached to a button instead of using `on_interaction`, that's what they're there for.

**Reply by ctrl-Felix** [reply]
> You are setting a custom_id explicitly, which might have a chance of conflict unless you're going out of way to ensure they are unique

They are definetly unique

>Any response you do after that is bound to fail

Shouldn't `interaction.followup.send` work?

Is there an example for the callback somewhere?

**Reply by Rapptz** [reply]
A followup isn't a response.

There are examples in the repository under the `views` folder.

**Comment by felipeyx** [comment]
The problem is that your global on_interaction listener is trying to acknowledge interactions that may already be handled by the View/button system:

@bot.listen()
async def on_interaction(interaction):
    await interaction.response.defer(...)

An interaction can only have one initial response. If the view callback already responded and the listener also tries to defer it, Discord returns:

40060: Interaction has already been acknowledged

I would remove the global defer and respond inside the component callback:

class MyView(discord.ui.View):
    @discord.ui.button(label="Click")
    async def button(
        self,
        interaction: discord.Interaction,
        button: discord.ui.Button,
    ):
        await interaction.response.defer(ephemeral=True)

        # do work...

        await interaction.followup.send("Done", ephemeral=True)

If you keep a global on_interaction, filter only the interaction types/custom IDs that it actually owns and check interaction.response.is_done() before trying to respond.
